### ISSUE TITLE
Implement WriteFilesExecTransformer discussion

### ISSUE THREAD

**Issue by JkSelf** [oap-project/gluten#3547]
### Description

Since Spark 3.4 introduced the WriteFilesExec operator to facilitate write operations in [SPARK-41708](https://issues.apache.org/jira/browse/SPARK-41708), we can utilize the WriteFilesExec operator to implement native Parquet write. This can be achieved by adding a WriteFilesExecTransformer and offloading the process to the Velox TableWriteNode. To achieve this, consider the following implementation steps that we can discuss together:
1. Add `WriteFilesRelNode`.
2. Add `WriteFilesExecTransformer`.
3. Implement `doExecuteWrite(writeFileSpec: WriteFileSpec): RDD[WriterCommitMessage]` in WholeStageTransformer.scala. Invoke the `doExecuteColumnar()` method in  in WholeStageTransformer.scala  to write the data. Capture the returned `RDD[ColumnBatch]`. Construct an `RDD[WriterCommitMessage]` by following these steps:

    - The `ColumnarBatch `returned by `TableWriteNode `contains `row`, `fragments`, `context `as below.
    - Enhance it by adding `partition `column and `stats `column.
    - Build the `ExecutedWriteSummary `object based on above stats information.
    - Finally, create a `WriterCommitMessage`.

Note: the current output layout in `TableWriteNode `is as below:
 ```
  // Page layout:
  // row     fragments     context    [partition]    [stats]
  // X         null          X        [null]          [null]
  // null       X            X        [null]          [null]
  // null       X            X        [null]          [null]
```
Do you have any input？@ulysses-you @zhztheplayer @zhouyuan

**Comment by ulysses-you** [comment]
thank you @JkSelf , 

For 2, the new added transformer should inherit the `WriteFilesExec`, an example in vanilla Spark is `ColumnarWriteExec`.

For 3, can you clarify a bit more what is the `fragments` and `context` ?

If I understand it correct, the new version of write data pipeline is, right ? We just need pass the iterator to native side and return write result back after all columnar batch are written.
<img width="1106" alt="image" src="https://github.com/oap-project/gluten/assets/12025282/b57b9a7f-bdd6-44cf-96d9-7b7a74bca4ba">

**Comment by Yohahaha** [comment]
One thoughts: could we do async write in native side, instead of in java side?

**Comment by baibaichen** [comment]
@binmahone FYI

**Comment by JkSelf** [comment]
Thanks for your prompt reply. @ulysses-you 
> 
> For 1, do we really need WriteFilesRelNode ? It seems rel node is only used when there is a mapping at native side. Do we want to add a write file native operator ?
>

We require WriteFilesRelNode to utilize the native TableWriterNode operator offered by Velox directly.
> 
> For 2, the new added transformer should inherit the `WriteFilesExec`, an example in vanilla Spark is `ColumnarWriteExec`.
>
My suggestion is to directly inherit `UnaryExecNode` instead of inheriting `WriteFilesExec`. This allows us to include `WriteFilesExecTransormer ` inside the `WholeStageTransformer`.
> For 3, can you clarify a bit more what is the `fragments` and `context` ?
> 
> If I understand it correct, the new version of write data pipeline is like below. We just need to pass the iterator handle to native side and return write result back after all columnar batch are written, right ? <img alt="image" width="1106" src="https://user-images.githubusercontent.com/12025282/278551231-b57b9a7f-bdd6-44cf-96d9-7b7a74bca4ba.png">

The fragement and the context are all some metric info. And the detailed information is here: [fragement ](https://github.com/facebookincubator/velox/blob/7ba1d402a2a697ddf761e049e00959a48ca7f320/velox/connectors/hive/HiveDataSink.cpp#L464) and [context](https://github.com/facebookincubator/velox/blob/7ba1d402a2a697ddf761e049e00959a48ca7f320/velox/exec/TableWriter.cpp#L236)

The native write pipeline is as below. 

![image](https://github.com/oap-project/gluten/assets/11972570/add2ff8d-330b-495d-bb8e-0b797f0a0c20)



Our  focus is to implement the doExecuteWrite method. By calling doExecuteColumnar, we offload the writing operator to Velox. The TableWriter operator in Velox handles the file writing operation and returns the RDD[ColumnarBatch]. This RDD contains the necessary information for creating the RDD[WriterCommitMessage].

**Comment by JkSelf** [comment]
> One thoughts: could we do async write in native side, instead of in java side?

@Yohahaha  Good point. Velox has now implemented the native operator TableWriteNode. To proceed, our initial step is to enable TableWriteNode and subsequently focus on optimizing asynchronous writing for performance improvement.

**Comment by ulysses-you** [comment]
@JkSelf OK, I got your point. However, we must make it inherit WriteFileExec due to there is a check in vanilla Spark. It's better to put the WriteFileExec node at the top level node of plan. Otherwise we should impelemnt executeWrite for WholeStageTransformer. So, I think the plan tree should be:

```
 ColumnarWriteFile (inherit WriteFile)
        |
 WholeStageTransformer
        |  
 WriteFileTransformer (mapping to native TableWriteNode)
        |
      query
```

I think the pipeline you point out is same with mine, and it would make confused... The native side can not return rdd, it should return something like WriteResult object.

**Comment by JkSelf** [comment]
@ulysses-you Indeed, with the introduction of the ColumnarWriteFilesExec operator, there is no need to implement the doExecuteWrite() method in the WholeStageTransformer. Instead, we can implement the doExecuteWrite() method within the ColumnarWriteFilesExec class itself, which improves code clarity.

**Comment by zhztheplayer** [comment]
> > One thoughts: could we do async write in native side, instead of in java side?
> 
> @Yohahaha Good point. Velox has now implemented the native operator TableWriteNode. To proceed, our initial step is to enable TableWriteNode and subsequently focus on optimizing asynchronous writing for performance improvement.

I think probably it's Velox's job to implement async write. The current async write code in Java was for the old Arrow parquet writer which was dropped long time ago. It was neither designed for performance purpose, nor for Velox's parquet writer.

**Comment by JkSelf** [comment]
@zhztheplayer  Yes, once we have support for WriteExecTransformer, we will remove the Java-side asynchronous write code. In the future, we may also consider implementing asynchronous writing support in the TableWrite operator of Velox, if necessary.
